### PR TITLE
Secure JCR folders

### PR DESCRIPTION
**Pull request by trongtt**

### CODE REVIEW

File: component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+  public SecureJCRFoldersUpgradePlugin(RepositoryService repoService, NodeHierarchyCreator nodeHierarchyCreator, InitParams initParams) {
+    super(initParams);
+    this.repoService = repoService;
+    this.nodeHierarchyCreator = nodeHierarchyCreator;
+  }
+
+  @Override
+  public boolean shouldProceedToUpgrade(String newVersion, String previousVersion) {
+    return VersionComparator.isAfter(newVersion,previousVersion);
```
Comment (thomasdelhomenie) [inline component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: Should not be required with the new version of the upgrade framework, by using the configuration plugin.upgrade.target.version.

File: component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    migrateCollaboration();
+
+    migrateGadgets();
+
+    migrateUsers();
+
+    migrateLoginHistory();
+  }
+
+  private void migrateCollaboration() {
```
Comment (thomasdelhomenie) [inline component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: Please add comments on each method to explain what it does.

File: component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import org.exoplatform.services.jcr.ext.common.SessionProvider;
+import org.exoplatform.services.jcr.ext.hierarchy.NodeHierarchyCreator;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+import org.exoplatform.services.security.IdentityConstants;
+
+import javax.jcr.Node;
+import javax.jcr.Session;
+
+public class SecureJCRFoldersUpgradePlugin extends UpgradeProductPlugin {
```
Comment (thomasdelhomenie) [inline component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: Please add unit tests for this migration.

Comment (trongtt) [reply component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: @thomasdelhomenie  does it mean we should test initializing data from the previous configuration, then perform this migration to validate it ?

Comment (thomasdelhomenie) [reply component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: I was talking about unit tests, not integration tests. Something like https://github.com/exo-addons/push-notifications/blob/develop/service/src/test/java/org/exoplatform/push/upgrade/PushNotificationSettingsUpgradePluginTest.java

Comment (trongtt) [reply component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: I do not find this is really useful in these cases indeed. It's mostly about removing permissions from nodes.

Can you re-consider if that's needed ? or give me some more hints ?

Comment (thomasdelhomenie) [reply component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: Unit tests are always useful. They will ensure we call the right permissions updates.

Comment (trongtt) [reply component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: Ok, I'm going to add some useful tests here.

Comment (trongtt) [reply component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: Done.

File: component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java [hunk 4]
```diff
@@ -30,18 +30,21 @@ public SecureJCRFoldersUpgradePlugin(RepositoryService repoService, NodeHierarch
...
 
-    migrateUsers();
+    migrateGadgets(sessionProvider);
 
-    migrateLoginHistory();
+    migrateUsers(sessionProvider);
+
+    migrateLoginHistory(sessionProvider);
+
+    sessionProvider.close();
```
Comment (thomasdelhomenie) [inline component/upgrade/plugins/src/main/java/org/exoplatform/platform/upgrade/plugins/SecureJCRFoldersUpgradePlugin.java]: Shouldn't we put this in a finally block to make sure it is always done ?

**Pull request merged**